projects/bootloader/src/main.c

```diff
@@ -1,21 +1,74 @@
+#include "bootloader_can.h"
+#include "bootloader_events.h"
+#include "can_datagram.h"
 #include "config.h"
 #include "crc32.h"
+#include "dispatcher.h"
+#include "event_queue.h"
 #include "flash.h"
 #include "interrupt.h"
 #include "jump_to_application.h"
 #include "log.h"
+#include "ping.h"
 #include "soft_timer.h"
+#include "wait.h"
+
+static CanStorage s_can_storage;
+static CanSettings s_can_settings = {
+  .loopback = false,
+  .bitrate = CAN_HW_BITRATE_500KBPS,
+  .rx_event = CAN_DATAGRAM_EVENT_RX,
+  .tx_event = CAN_DATAGRAM_EVENT_TX,
+  .fault_event = CAN_DATAGRAM_EVENT_FAULT,
+  .tx = { GPIO_PORT_A, 12 },
+  .rx = { GPIO_PORT_A, 11 },
+};
+
+static CanDatagramSettings s_datagram_settings = {
+  .tx_event = DATAGRAM_EVENT_TX,
+  .rx_event = DATAGRAM_EVENT_RX,
+  .repeat_event = DATAGRAM_EVENT_REPEAT,
+  .error_event = DATAGRAM_EVENT_ERROR,
+  .error_cb = NULL,
+};
 
 int main(void) {
   LOG_DEBUG("Hello from the bootloader!\n");
-  // initialize all the modules
+
   flash_init();
   interrupt_init();
   soft_timer_init();
   crc32_init();
+  event_queue_init();
   config_init();
 
+  // gets the board id
+  BootloaderConfig blconfig = {
+    .controller_board_id = 0xFF,
+  };
+  config_get(&blconfig);
+  uint8_t board_id = blconfig.controller_board_id;
+
+  // can and datagram
+  bootloader_can_init(&s_can_storage, &s_can_settings, board_id);
+  can_datagram_init(&s_datagram_settings);
+
+  dispatcher_init(board_id);
+  ping_init(board_id);
+
   jump_to_application();
```

Comment:
Might be a good Idea to add a jump_to_application_init() function similar to ping_init, that can be added to the initial part of the main function and will call jump_to_application() from its own library, abstracting out the details from the main loop.

Would be good to have @ryandancy 's thoughts

Reply:
The `jump_to_application()` call will actually jump to the application, it won't return, so in order to actually process commands that call can't be there. Once the jump-to-application command module is complete it'll use the `jump_to_application` to actually do the jumping to application - it's probably good to take out the call right now so it's possible to test ping? It doesn't really matter though.

Reply:
it doesn't matter for my testing, but if you need to test it on an actual board, it should probably be removed. I'll just register it to its datagram command with dispatcher, so it can still be called if it's needed. There should probably be a init function for that, but it's not in my ticket :P